modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,489 @@
+    (date-time->substitution (.format ts DateTimeFormatter/ISO_OFFSET_DATE_TIME))))
+
+(defmethod sql.params.substitution/->prepared-substitution [:presto-jdbc LocalDateTime]
+  [_ ^LocalDateTime t]
+  ;; similar to above implementation, but for `LocalDateTime`
+  ;; when Presto parses this, it will account for session (report) time zone
+  (date-time->substitution (.format t DateTimeFormatter/ISO_LOCAL_DATE_TIME)))
+
+(defmethod sql.params.substitution/->prepared-substitution [:presto-jdbc OffsetDateTime]
+  [_ ^OffsetDateTime t]
+  ;; similar to above implementation, but for `ZonedDateTime`
+  ;; when Presto parses this, it will account for session (report) time zone
+  (date-time->substitution (.format t DateTimeFormatter/ISO_OFFSET_DATE_TIME)))
+
+(defn- set-time-param
+  "Converts the given instance of `java.time.temporal`, assumed to be a time (either `LocalTime` or `OffsetTime`)
+  into a `java.sql.Time`, including milliseconds, and sets the result as a parameter of the `PreparedStatement` `ps`
+  at index `i`."
+  [^PreparedStatement ps ^Integer i ^Temporal t]
+  ;; for some reason, `java-time` can't handle passing millis to java.sql.Time, so this is the most straightforward way
+  ;; I could find to do it
+  ;; reported as https://github.com/dm3/clojure.java-time/issues/74
+  (let [millis-of-day (.get t ChronoField/MILLI_OF_DAY)]
+    (.setTime ps i (Time. millis-of-day))))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetTime]
+  [_ ^PreparedStatement ps ^Integer i t]
+  ;; necessary because `PrestoPreparedStatement` does not implement the `setTime` overload having the final `Calendar`
+  ;; param
+  (let [adjusted-tz (t/with-offset-same-instant t (t/zone-offset 0))]
+    (set-time-param ps i adjusted-tz)))
+
+(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc LocalTime]
+  [_ ^PreparedStatement ps ^Integer i t]
+  ;; same rationale as above
+  (set-time-param ps i t))
+
+(defn- ^LocalTime sql-time->local-time
+  "Converts the given instance of `java.sql.Time` into a `java.time.LocalTime`, including milliseconds. Needed for
+  similar reasons as `set-time-param` above."
+  [^Time sql-time]
+  ;; Java 11 adds a simpler `ofInstant` method, but since we need to run on JDK 8, we can't use it
+  ;; https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalTime.html#ofInstant(java.time.Instant,java.time.ZoneId)
+  (let [^LocalTime lt (t/local-time sql-time)
+        ^Long millis  (mod (.getTime sql-time) 1000)]
+    (.with lt ChronoField/MILLI_OF_SECOND millis)))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:presto-jdbc Types/TIME]
+  [_ ^ResultSet rs ^ResultSetMetaData rs-meta ^Integer i]
+  (let [type-name  (.getColumnTypeName rs-meta i)
+        base-type  (presto-common/presto-type->base-type type-name)
+        with-tz?   (isa? base-type :type/TimeWithTZ)]
+    #(let [local-time (-> (.getTime rs i)
```

Comment:
small nit but not a fan of returning a thunk like this. very easy to miss that it's a function rather than just a regular expression. Seems like all the others use `(fn []` and some even have names

Reply:
Fixed.